### DISCUSSION TITLE
Step-by-Step Tutorial for Installing Outline on Digital Ocean with Ubuntu 22.04

### DISCUSSION THREAD

**Discussion by thomasgriffin** [outline/outline#3668] [Show and tell]
Hey everyone,

I see lots of people have experienced many of the same issues that I myself experienced trying to setup Outline in a production environment. After spending countless hours off and on over the last 6 months, I've finally completed a [step-by-step tutorial for installing Outline on Digital Ocean with Ubuntu 22.04](https://thomasgriffin.com/how-to-install-the-outline-knowledge-base-wiki-on-ubuntu/).

I've gone through the tutorial 4 times now to validate the results, and I can confirm they are repeatable. I've done my best to include helpful screenshots and commentary, and I've also implemented some best practices to secure the installation as well.

**Much of the stuff I have in the tutorial is either not listed in the official docs at all or simply assumed in the docs.** For example, in the sample `.env` file, the URL connections for Postgres and Redis simply do not work when trying to put this in production. Or did you know that if you don't put the `POSTGRES_DB` variable before the `POSTGRES_USER` and `POSTGRES_PASS` variables, it won't create the database and roles correctly in Docker? Additionally, some things are more complex than they need to be, and where I have experienced that, I've done my best to simplify the process.

Outline is awesome. I love it. My hope is that this tutorial helps others install it successfully so that they can experience using it with their teams too. 💯

You can find the tutorial here: https://thomasgriffin.com/how-to-install-the-outline-knowledge-base-wiki-on-ubuntu/

Feel free to drop comments there or here, including any feedback about the tutorial or snags you run into along the way.

**Comment by ReessKennedy** [comment]
Amazing tutorial here @thomasgriffin ! I had an existing server running Docker, Redis and Postgres with other stuff installed on the same server but used your tutorial to understand how to hook them together and it was especially useful re: the AWS S3 settings which don't seem very detailed in the default install instructions. Well done here, Thomas. Thank you.

**Comment by thomasgriffin** [comment]
The feedback on the tutorial has been amazing! I've seen and heard about so many people getting Outline setup properly for the first time by following it.

I wanted to drop back here and mention a couple of updates I have made to the tutorial, specifically:

- Fixing the monthly cron so that it works properly. It needed an executable command, and I fixed a couple of docker commands to ensure that it pulled down the latest image of Outline.
- Fixing an issue with AWS bucket authentication, changing the ACLs to enabled and toggling the Bucket owner preferred setting on.
- Fixing an issue with 403 access denied requests for assets uploaded to S3. The blanket "Block public access" setting being toggled on was no good, so turning it off fixed those issues.

Anyways, hope it continues to be helpful!

**Comment by thomasgriffin** [comment]
I've made a couple of additional updates to the tutorial, including fixing an issue where a deprecated yarn command has been updated with the new command, as well as adding support for daily cron pings that Outline needs to trigger various services.

I'll be testing the tutorial again this month to make sure that things are updated where needed to support WebSockets editing.

And, I'm delighted to share that more than 1,000 people have successfully set up Outline using the tutorial above. 😎

**Comment by aresstokrat** [comment]
Demo site is not working. Trying to login with simple email sign up h it doesn't send email with activation to me.

**Reply by tommoor** [reply]
Not really sure what you're referring to for a demo site, the app is hosted here: https://app.getoutline.com/

**Reply by aresstokrat** [reply]
Your site is Impossible to login with email.
Only "special" emails for workgroups, I'm just an ordinary guy with a bunch of friends, I don't have any special emails. 


But I'm talking about https://outline.thomasgriffin.com/

**Reply by thomasgriffin** [reply]
@aresstokrat There is an existing feature request for this (local authentication) that has yet to be implemented: https://github.com/outline/outline/issues/1881

**Reply by aresstokrat** [reply]
> @aresstokrat There is an existing feature request for this (local authentication) that has yet to be implemented: #1881

I'm not into local auth.
I'm into smtp login, like at your demo site (but it's not sending password actually, I think you didn't config smtp)

**Reply by aresstokrat** [reply]
> @aresstokrat There is an existing feature request for this (local authentication) that has yet to be implemented: #1881

Ps. Sent you a message from your site about a week ago 🙏

**Reply by thomasgriffin** [reply]
@aresstokrat I think what you are looking for is to sign up for an account on the Outline hosted platform at getoutline.com. That will give you a demo of the system itself, and then you can determine if you want to install the open source software or not.

**Reply by aresstokrat** [reply]
That's the first thing I tryed to do. Impossible. They asking for google , slack, Microsoft workspace corp accounts for sign up, I don't have corp accounts.  I'm just an ordinary guy )

**Reply by tommoor** [reply]
There are a plethora of products designed for individual note taking, Outline is designed for teams hence the available sign-in methods.

**Reply by aresstokrat** [reply]
Me and my teammates don't use corp accounts. Why should we? We are independent individuals. Not corp workers. Community with a hobby. Still we are making projects together.
For individual work we use obsidian.

**Reply by Limezy** [reply]
You can open a slack workspace for free, then log in Outline, then activate the login by email through magic link for each of you. No need to use that slack account at all. As said in another post, you can alternatively install Yunohost as your community server management system, Outline is available as an "app" for one click installation, and you can then use the Yunohost local user base for authentication. I'm sure there are some other alternatives too !

**Comment by thomasgriffin** [comment]
I've just completed walking through the tutorial again to validate everything is working as expected. All is good!

Best to everyone using it to get Outline up and running on their own server. 🚀

**Reply by ericklindberg** [reply]
I don't really comment on too many things here, however, I just wanted to say thank you! Your guide was clear enough to allow anybody to manipulate it well enough to get it working. By that, I mean that I was able to get it set up with a MiniIO "S3" and Slack authorization, using HA Proxy as the proxy. Again, really appreciate it! I'll gladly share my notes if anybody needs help with this.

**Reply by thomasgriffin** [reply]
Awesome, glad to hear it was helpful for you!

**Comment by huds0nx** [comment]
I finished your tutorial, it was great, but when finally trying to access the domain, I am getting an nginx "502 Bad Gateway" error. When I check my docker logs, I see these two errors coming up frequently:
`recv() failed (104: Connection reset by peer) while reading response header from upstream, client: {my ip}, server: {my domain}, request: "GET /favicon.ico HTTP/1.1", upstream: "http://127.0.0.1:3000/favicon.ico", host: {my droplet ip}`

`upstream prematurely closed connection while reading response header from upstream, client: {my ip}, server: {my domain}, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: {my droplet ip}`

Any ideas? Help would mean the world :)

**Comment by thomasgriffin** [comment]
@huds0nx Can you try setting `user: root` in your `docker-compose.yml` file? I have heard a few people say that has worked for them!

**Reply by huds0nx** [reply]
Thanks for your reply, I tried that and unfortunately I am still getting the errors. Do you have an email I can best contact you at? I don’t think the contact form on your site is working.

**Reply by thomasgriffin** [reply]
@huds0nx I would also try changing the ownership of the `/opt/outline` directory to be owned by the `nodejs` user, as that has also been mentioned as a possible fix for that 502 error.

Also, thank you for letting me know about the contact form being broken! I have gotten it fixed (bad caching), so it should work now. Feel free to use that to reach out directly.

**Reply by terencehangtight** [reply]
> @huds0nx I would also try changing the ownership of the `/opt/outline` directory to be owned by the `nodejs` user, as that has also been mentioned as a possible fix for that 502 error.
> 
> Also, thank you for letting me know about the contact form being broken! I have gotten it fixed (bad caching), so it should work now. Feel free to use that to reach out directly.

Hey Thomas - great guide... running into the same issues. I notice Outline hanging up when looking at existing docker images.

<img width="1294" alt="Screenshot 2025-07-03 at 11 13 53 AM" src="https://github.com/user-attachments/assets/01a78e84-ec1d-4032-b975-f4fa84c5d259" />

**Reply by terencehangtight** [reply]
After a few more checks... Outline keep restarting and looking at the logs, it looks like Yarn can't get kicked up. 

<img width="1355" alt="Screenshot 2025-07-03 at 11 20 27 AM" src="https://github.com/user-attachments/assets/fcdeb448-4e4c-4c24-90f3-9ae7daf21fc7" />

**Reply by terencehangtight** [reply]
Solved the process being killed. Not enough memory!

**Comment by thomasgriffin** [comment]
I've completed another round of updates to the tutorial, and I've also added support for the new local file storage feature from `v0.72.0` and onward.

All the best as you cross the finish line of 2023 and step confidently into 2024!

**Comment by AshwinPrasadRao** [comment]
Thank you for this fantastic guide, @thomasgriffin! I am a beginner to all this and followed your guide to the tee. 

However, I also face the 502 Gateway error after doing everything (except adding SMTP). Docker Logs show no errors. In the Nginx error logs file, I saw the same kind of errors as @huds0nx reported above. 

Setting `user: root` in your `docker-compose.yml` file did not help. Restarting docker containers with `docker-compose down --remove-orphans && docker-compose up -d` did not help either. I would really appreciate any help you can give me here.

**Reply by thomasgriffin** [reply]
Thanks for the kind words! Did you also try changing the ownership of the `/opt/outline` directory to be owned by the `nodejs` user?

**Reply by AshwinPrasadRao** [reply]
Thank you for responding, @thomasgriffin. I’m a beginner, and this set-up is for a nonprofit that does not have an IT team to run these things. 

To ensure that I don’t break anything by accident as I go about changing the ownership of the /opt/outline directory - the command would be: 

`Chown thomas /opt/outline `

Right? 

Thanks again, and my apologies for asking you about rudimentary things.

**Comment by thomasgriffin** [comment]
You can do chown -R {user} /opt/outline :)

---------------------
Thomas Griffin
Advisor
Awesome Motive Inc.
Website <https://thomasgriffin.com> | LinkedIn
<https://www.linkedin.com/in/jthomasgriffin>
​
​This email transaction is considered private and should not be shared
unless otherwise noted.​

**Comment by smegoff** [comment]
Hello!

Great tutorial - however I am still stuck on a bad gateway.

For testing I am using VMWare workstation and Ubuntu Server 24.04

2024/11/07 03:30:12 [error] 11031#11031: *15 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 192.168.101.254, server: docs.engagetech.nz, request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:3000/", host: "docs.engagetech.nz"
2024/11/07 03:30:12 [error] 11031#11031: *15 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 192.168.101.254, server: docs.engagetech.nz, request: "GET /favicon.ico HTTP/1.1", upstream: "http://127.0.0.1:3000/favicon.ico", host: "docs.engagetech.nz", referrer: "https://docs.engagetech.nz/"
2024/11/07 03:30:12 [error] 11031#11031: *19 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 192.168.101.254, server: docs.engagetech.nz, request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:3000/", host: "docs.engagetech.nz"
2024/11/07 03:30:12 [error] 11031#11031: *19 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 192.168.101.254, server: docs.engagetech.nz, request: "GET /favicon.ico HTTP/1.1", upstream: "http://127.0.0.1:3000/favicon.ico", host: "docs.engagetech.nz", referrer: "https://docs.engagetech.nz/"

I have tried all of the above suggestions with an exception where I dont have an /opt/outline directory created, mine is /opt/containerd/.

This is the second time I have attempted to stand this up now with the same result. 

Perhaps the tutorial isnt supported for 24.04 yet?

Cheers

Sean

**Comment by smegoff** [comment]
So I just tried 22.04 with the same result. 

Not sure what I am doing wrong :-/

**Comment by smegoff** [comment]
well, I came in this morning, and it was working. Now to set up Azure Entra Auth. I guess I was being impatient!

**Reply by thomasgriffin** [reply]
Awesome, glad to hear it worked out! I'll be reviewing the tutorial again and running through it in the next couple of weeks to make sure all latest updates are working properly.

**Comment by smegoff** [comment]
Incidentally, what's the process for pulling the latest image with your method of install (I used your tutorial to install it self hosted)? I attempted to do it manually, however it broke my installation, so have to restore from a snapshot.

I simply did this from the docs:

Updating

To update Outline, follow these steps:
[Backup the database](https://github.com/s/hosting/doc/backups-KZtPOADCHG)
Update to the latest image
Use docker pull docker.getoutline.com/outlinewiki/outline:latest to download the image or
If you are using docker-compose with a fixed version number then you will need to update the docker-compose.yml file to point to the latest version.

Run the container

**Comment by smegoff** [comment]
Ok - do It didnt break it in the end, just docker takes ages to load the first time - there are 5 packages it tries to download and build. 

So it would seem now pulling the lastest image doesn't bring it up to the latest version. 

https://github.com/outline/outline/releases/tag/v0.81.0

After running the pull.

![image](https://github.com/user-attachments/assets/2310c42d-d0c1-4bba-a4d3-ea56f89d89d3)

**Comment by dzhulianaN** [comment]
@thomasgriffin Man, that's super cool - very much appreciated your hard work! It's July 2025, and the guide still works with all services! Faced few errors but nothing dramatic, setup the project within few hours - you saved me a lot of time, thanks!
- Dzhuliana

**Reply by thomasgriffin** [reply]
That's fantastic to hear! I do plan to get around to reviewing and updating it at some point this year to make sure it is working smoothly with the latest and greatest updates. 😊
